### ISSUE TITLE
The result of test.py is different from detect.py

### ISSUE THREAD

**Issue by Li-bing0114** [ultralytics/yolov5#3130]
Both test.py and detect.py can save the txt file of prediction box. Why is the test result better than the inference one？  Thank you for your reply

**Comment by Li-bing0114** [comment]
Input the same picture into the test.py and detect.py separately，and different results are obtained（different predicted box）

**Comment by fcakyon** [comment]
@Li-bing0114 thresholds are different https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/test.py#L295
https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/detect.py#L154

**Comment by Li-bing0114** [comment]
The conf_threshold and iou_threshold in test.py and detect.py  were changed to the same (0.001 and 0.6 )，and the model is the same，but the different results are obtained

**Comment by fcakyon** [comment]
nms inputs are also different in detect and test scripts:
https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/detect.py#L71
https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/test.py#L122

**Comment by glenn-jocher** [comment]
@Li-bing0114 yes as @fcakyon these two files use different default settings and have different purposes. More details on the differences between [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py), [detect.py](https://github.com/ultralytics/yolov5/blob/master/detect.py) and [test.py](https://github.com/ultralytics/yolov5/blob/master/test.py) in YOLOv5 are blow.

These 3 files are designed for different purposes and utilize different dataloaders with different settings. train.py dataloaders are designed for a speed-accuracy compromise, test.py is designed to obtain the best mAP on a validation dataset, and detect.py is designed for best real-world inference results. A few important aspects of each:

## `train.py`

- **trainloader** LoadImagesAndLabels(): designed to load train dataset images and labels. Augmentation capable and enabled. https://github.com/ultralytics/yolov5/blob/fca5e2a48fb526b57bda0c66be6b7ac1aaa8d83d/train.py#L188-L192
- **testloader** LoadImagesAndLabels(): designed to load val dataset images and labels. Augmentation capable but disabled. https://github.com/ultralytics/yolov5/blob/fca5e2a48fb526b57bda0c66be6b7ac1aaa8d83d/train.py#L199-L202
- image size: `640`
- rectangular inference: `False`
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L291): `0.001`
- [iou threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L292): `0.6`
- [multi-label](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L120): `True`
- [padding](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L83-L90): `None`

## `test.py`

- **dataloader** LoadImagesAndLabels(): designed to load train, val, test dataset images and labels. Augmentation capable but disabled. https://github.com/ultralytics/yolov5/blob/fca5e2a48fb526b57bda0c66be6b7ac1aaa8d83d/test.py#L89-L90
- image size: `640`
- rectangular inference: `True`
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L291): `0.001`
- [iou threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L292): `0.6`
- [multi-label](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L120): `True`
- [padding](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L83-L90): `0.5 * maximum stride`

## `detect.py`

- **dataloaders** (multiple): designed for loading multiple types of media (images, videos, globs, directories, streams). https://github.com/ultralytics/yolov5/blob/fca5e2a48fb526b57bda0c66be6b7ac1aaa8d83d/detect.py#L46-L53
- image size: `640`
- rectangular inference: `True`
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/detect.py#L152): `0.25` 
- [iou threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/detect.py#L153): `0.45`
- [multi-label](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/utils/general.py#L415): False
- [padding](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/detect.py#L45-L54): `None`

## YOLOv5 PyTorch Hub

`models.autoShape()` class used for image loading, preprocessing, inference and NMS. For more info see [YOLOv5 PyTorch Hub Tutorial](https://docs.ultralytics.com/yolov5/tutorials/pytorch_hub_model_loading) https://github.com/ultralytics/yolov5/blob/fca5e2a48fb526b57bda0c66be6b7ac1aaa8d83d/models/common.py#L225-L250

- image size: `640`
- rectangular inference: `True`
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/models/common.py#L226-L230): `0.25 `
- [iou threshold](https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/models/common.py#L226-L230): `0.45`
- [multi-label](https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/models/common.py#L288): False
- [padding](https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/models/common.py#L226-L230): `None`

**Comment by fcakyon** [comment]
@glenn-jocher how does `multi-label` parameter affects the results?

**Comment by glenn-jocher** [comment]
@fcakyon multi-label is an NMS parameter that allows for multiple classes to be detected on a single box:
https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/utils/general.py#L461-L468

**Comment by fcakyon** [comment]
@glenn-jocher  i see, thanks a lot 👍

**Comment by keivanmoazami** [comment]
in test.py ground truth labels pass to nms and then nms function add ground truth labels to final result. why ????
in utils/general.py line 524
```
if labels and len(labels[xi]):
    l = labels[xi]
    v = torch.zeros((len(l), nc + 5), device=x.device)
    v[:, :4] = l[:, 1:5]  # box
    v[:, 4] = 1.0  # conf
    v[range(len(l)), l[:, 0].long() + 5] = 1.0  # cls
    x = torch.cat((x, v), 0)
```
and this change final result in test.py. but labels not exist in detect.py
just remove labels parameter from test.py line 122

**Comment by glenn-jocher** [comment]
@keivanmoazami this code is used when hybrid-autolabelling is selected in test.py:
```
python test.py --save-hybrid
```
https://github.com/ultralytics/yolov5/blob/f3c3d2ce5d85ba77336a9d0a87c6a446732cdda6/test.py#L302-L305
